### ISSUE TITLE
New rule: Always inject dispatchers

### ISSUE THREAD

**Issue by chao2zhang** [detekt/detekt#4175]
## Expected Behavior of the rule
When standard dispatchers like `Dispatchers.IO/Default/Unconfined` are being used directly without dependency injection, violations should be reported.

The following scenarios shouldn't be reported.
- Coroutine dispatchers used as a default argument of the constructor.
- `Dispatchers.Main`/`Dispatchers.Main.Immediate` because this is overrideable by [Dispatchers.setMain] (https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-test/kotlinx.coroutines.test/set-main.html) and [Dispatchers.resetMain](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-test/kotlinx.coroutines.test/reset-main.html)

## Context
If a coroutine dispatcher is not injected, it is extremely difficult to write unit tests that can complete.
This is recommended by https://developer.android.com/kotlin/coroutines/coroutines-best-practices#inject-dispatchers
as well as https://youtu.be/KMb0Fs8rCRs?t=839.

**Comment by marschwar** [comment]
Playing devils advocate here: Should this be a part of an android rule set  instead of the detekt core rule sets?

**Comment by cortinico** [comment]
> Playing devils advocate here: Should this be a part of an android rule set instead of the detekt core rule sets?

Not really. `Dispatchers.IO/Default/Unconfined` could be used also in non-Android project (i.e. backend) and follow the same guideline imho

**Comment by grndvl1** [comment]
Problem with this rule is in koin you need to declare what dispatcher you are injecting.  This then violates the rule.  I've been shutting off this cause its annoying.

**Comment by jillesme** [comment]
Hello, 

I was wondering if this rule is still recommended when using it in combination with Spring's Dependency Injection? It won't be able to find a `Bean` with `CoroutineDispatch`:

> "Could not autowire. No beans of 'CoroutineDispatcher' type found. " 
![IntelliJ IDEA  2024-10-21 at 14 21 32@2x](https://github.com/user-attachments/assets/04b839b4-4f0e-4db9-8a92-e39c2afb9df0)

This "error" makes sense, as we don't have a `CoroutineDispatcher` bean anywhere. Nor do we want that. We can fix the error by making it not required in Spring:  

```kt
@Autowired(required = false) private val dispatcher: CoroutineDispatcher = Dispatchers.IO,
```

However, the question remains whether it is a _best practice_ in this situation. If someone declared a bean as follows:

```kt
@Bean
fun oops() = Dispatchers.Main
```

Then all dependency injected dispatchers are ran on the main thread instead of their default assigned dispatcher. Of course one would say "well don't declare a Spring Bean of type CoroutineDispatcher" which is fair, but that is harder to enforce in every Kotlin/Spring repository. Would love to hear your thoughts!
